openfl/utilities/fed_timer.py

```diff
@@ -0,0 +1,191 @@
+        if task.is_alive():
+            raise TimeoutError(f"Timeout after {self._max_timeout} second(s), Exception method: ({self._fn_name})") # noqa
+
+        return task.result()
+
+
+class SyncAsyncTaskDecoFactory:
+    '''
+    `Sync` and `Async` Task decorator factory allows creation of
+    concrete implementation of `wrapper` interface and `contextmanager` to
+    setup a common functionality/resources shared by `async_wrapper` and `sync_wrapper`.
+
+    '''
+
+    @contextmanager
+    def wrapper(self, func, *args, **kwargs):
+        yield
+
+    def __call__(self, func):
+        '''
+        Call to `@fedtiming()` executes `__call__()` method
+        delegated from the derived class `fedtiming` implementing `SyncAsyncTaskDecoFactory`.
+        '''
+
+        # Closures
+        self.is_coroutine = asyncio.iscoroutinefunction(func)
+        str_fmt = "{} Method ({}); Co-routine {}"
+
+        @wraps(func)
+        def sync_wrapper(*args, **kwargs):
+            '''
+            Wrapper for synchronous execution of decorated function.
+            '''
+            logger.debug(str_fmt.format("sync", func.__name__, self.is_coroutine))
+            with self.wrapper(func, *args, **kwargs):
+                return self.task.sync_execute()
+
+        @wraps(func)
+        async def async_wrapper(*args, **kwargs):
+            '''
+            Wrapper for asynchronous execution of decorated function.
+            '''
+            logger.debug(str_fmt.format("async", func.__name__, self.is_coroutine))
+            with self.wrapper(func, *args, **kwargs):
+                return await self.task.async_execute()
+
+        # Identify if the decorated function is `async` or `sync` and return appropriate wrapper.
+        if self.is_coroutine:
+            return async_wrapper
+        return sync_wrapper
+
+
+class fedtiming(SyncAsyncTaskDecoFactory):
+    def __init__(self, timeout):
+        self.timeout = timeout
+
+    @contextmanager
+    def wrapper(self, func, *args, **kwargs):
+        '''
+        Concrete implementation of setup and teardown logic, yields the control back to
+        `async_wrapper` or `sync_wrapper` function call.
+
+        Raises:
+            Exception: Captures the exception raised by `async_wrapper`
+            or `sync_wrapper` and terminates the execution.
+        '''
+        self.task = PrepareTask(
+            target_fn=func,
+            timeout=self.timeout,
+            args=args,
+            kwargs=kwargs
+        )
+        try:
+            start = time.perf_counter()
+            yield
+            logger.info(f"({self.task._fn_name}) Elapsed Time: {time.perf_counter() - start}")
+        except Exception as e:
+            template = "An exception of type {0} occurred. Arguments:\n{1!r}"
+            message = template.format(type(e).__name__, e.args[0])
+            logger.exception(message)
+            time.sleep(2)
```

Comment:
It is better two use named variables instead magic numbers. It is not obvious why is this timeout needed?

Reply:
Timeout is not really needed. I used it to just have a delay before the SIGKILL kicks in.
This functionality would be extended later to support graceful termination and this particular os._exit() call is not ideal.